### PR TITLE
New encoding API

### PR DESCRIPTION
**Pull request by hamishun**
Victor's new encoding APIs. Introducing the new `PtxtArray` and `EncodedPtxt` objects.

### CODE REVIEW

File: include/helib/EncodedPtxt.h [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+#define HELIB_ENCODED_PTXT_H
+
+#include <helib/DoubleCRT.h>
+#include <helib/norms.h>
+
+namespace helib {
+
+
+
+
```
Comment (hamishun) [inline include/helib/EncodedPtxt.h:22]:
```suggestion
```

File: include/helib/EncodedPtxt.h [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+ */
+#ifndef HELIB_ENCODED_PTXT_H
+#define HELIB_ENCODED_PTXT_H
+
+#include <helib/DoubleCRT.h>
+#include <helib/norms.h>
+
+namespace helib {
+
+
```
Comment (hamishun) [inline include/helib/EncodedPtxt.h:20]:
```suggestion
```

File: include/helib/EncodedPtxt.h [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+
```
Comment (hamishun) [inline include/helib/EncodedPtxt.h:21]:
```suggestion
```

File: include/helib/EncodedPtxt.h (hunk 1 above)
Comment (hamishun) [inline include/helib/EncodedPtxt.h:22]:
```suggestion
```

File: include/helib/EncodedPtxt.h [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+
+
```
Comment (hamishun) [inline include/helib/EncodedPtxt.h:23]:
```suggestion
```

File: include/helib/EncodedPtxt.h [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+  }
+
+  void reset() 
+  {
+    rep.set_ptr(0);
+  }
+
+};
+
+
```
Comment (hamishun) [inline include/helib/EncodedPtxt.h:176]:
```suggestion
```

File: include/helib/EncodedPtxt.h [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+
```
Comment (hamishun) [inline include/helib/EncodedPtxt.h:177]:
```suggestion
```

File: include/helib/EncodedPtxt.h [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+};
+
+class FatEncodedPtxt_base {
+public:
+
+  virtual ~FatEncodedPtxt_base() {}
+
+  virtual FatEncodedPtxt_base* clone() const = 0;
+  // makes this usable with cloned_ptr
```
Comment (hamishun) [inline include/helib/EncodedPtxt.h:242]:
```suggestion
```

File: include/helib/EncodedPtxt.h [hunk 8]
```diff
@@ -0,0 +1,343 @@
...
+    : dcrt(dcrt_), mag(mag_), scale(scale_), err(err_)
+  { }
+
+};
+
+class FatEncodedPtxt_base {
+public:
+
+  virtual ~FatEncodedPtxt_base() {}
+
```
Comment (hamishun) [inline include/helib/EncodedPtxt.h:240]:
```suggestion

 // TODO make this usable with cloned_ptr
```

File: include/helib/EncryptedArray.h [hunk 9]
```diff
@@ -73,8 +75,13 @@ typedef std::complex<double> cx_double;
   tab.restoreContext();
 
 class PlaintextArray; // forward reference
```
Comment (hamishun) [inline include/helib/EncryptedArray.h:77]: If PtxtArray takes over form PlaintextArray should we delete any refs to the latter?

File: include/helib/EncryptedArray.h [hunk 10]
```diff
@@ -968,10 +1028,52 @@ class EncryptedArrayCx : public EncryptedArrayBase
   {
     zzX tmp;
     double f = encode(tmp, pt, useThisSize, precision);
-    ::helib::convert(ptxt, tmp);
+    convert(ptxt, tmp);
     return f;
   }
 
+  //========= new EncodedPtxt interface
```
Comment (hamishun) [inline include/helib/EncryptedArray.h:1035]:
```suggestion
  // New EncodedPtxt interface
```

File: include/helib/EncryptedArray.h [hunk 11]
```diff
@@ -1504,6 +1817,281 @@ void applyPerm(const EncryptedArray& ea,
...
+{
+  rotate1D(a.ea, a.pa, i, k);
+}
+
+inline void shift1D(PtxtArray& a, long i, long k)
+{
+  shift1D(a.ea, a.pa, i, k);
+}
+
+
```
Comment (hamishun) [inline include/helib/EncryptedArray.h:2005]:
```suggestion
```

File: include/helib/EncryptedArray.h [hunk 12]
```diff
@@ -1504,6 +1817,281 @@ void applyPerm(const EncryptedArray& ea,
...
+
```
Comment (hamishun) [inline include/helib/EncryptedArray.h:2006]:
```suggestion
```

File: include/helib/EncryptedArray.h [hunk 13]
```diff
@@ -1504,6 +1817,281 @@ void applyPerm(const EncryptedArray& ea,
...
+inline void frobeniusAutomorph(PtxtArray& a, const NTL::Vec<long>& vec)
+{ frobeniusAutomorph(a.ea, a.pa, vec); }
+
+inline void applyPerm(PtxtArray& a, const NTL::Vec<long>& pi)
+{ applyPerm(a.ea, a.pa, pi); }
+
+inline void power(PtxtArray& a, long e)
+{ power(a.ea, a.pa, e); }
+
+
```
Comment (hamishun) [inline include/helib/EncryptedArray.h:2089]:
```suggestion
```

File: include/helib/EncryptedArray.h [hunk 14]
```diff
@@ -663,65 +768,40 @@ class EncryptedArrayDerived : public EncryptedArrayBase
...
   const PAlgebraModCx& alMod;
+
   zzX iEncoded; // an encoded plaintext with i in all the slots
+  // VJS-FIXME: this is a bad idea...see commenets in EaCx.cpp
 
 public:
   static double roundedSize(double x)
   {
-    long rounded = ceil(fabs(x));
+    long rounded = std::ceil(fabs(x));
```
Comment (hamishun) [inline include/helib/EncryptedArray.h:783]:
```suggestion
    long rounded = std::ceil(std::fabs(x));
```

File: tests/TestMatmulCKKS.cpp [hunk 15]
```diff
@@ -0,0 +1,222 @@
...
+  }
+
+  virtual void TearDown() override
+  {
+    if (helib_test::verbose) {
+      helib::printAllTimers();
+#if (defined(__unix__) || defined(__unix) || defined(unix))
+      struct rusage rusage;
+      getrusage(RUSAGE_SELF, &rusage);
+      cout << "  rusage.ru_maxrss=" << rusage.ru_maxrss << endl;
```
Comment (esteffin) [inline tests/TestMatmulCKKS.cpp:160]:
```suggestion
      std::cout << "  rusage.ru_maxrss=" << rusage.ru_maxrss << std::endl;
```

File: misc/legacy_tests/Test_CKKS_matmul.cpp [hunk 16]
```diff
@@ -0,0 +1,190 @@
...
+
+  std::cout << w << "\n";
+  std::cout << w1 << "\n";
+
+  if (verbose) {
+    printAllTimers(cout);
+#if (defined(__unix__) || defined(__unix) || defined(unix))
+      struct rusage rusage;
+      getrusage( RUSAGE_SELF, &rusage );
+      cout << "  rusage.ru_maxrss="<<rusage.ru_maxrss << endl;
```
Comment (esteffin) [inline misc/legacy_tests/Test_CKKS_matmul.cpp:138]:
```suggestion
      std::cout << "  rusage.ru_maxrss="<<rusage.ru_maxrss << std::endl;
```

File: misc/legacy_tests/Test_CKKS_matmul.cpp [hunk 17]
```diff
@@ -0,0 +1,190 @@
...
+  amap.arg("force_hoist", fhe_test_force_hoist, 
+           "-1 to force off"); 
+  amap.arg("ks_strategy", ks_strategy,
+           "0: default, 1:full, 2:bsgs, 3:minimal"); 
+
+
+  amap.parse(argc, argv);
+
+  if (verbose) {
+    cout << "*** Test_MatMul: m=" << m
```
Comment (esteffin) [inline misc/legacy_tests/Test_CKKS_matmul.cpp:171]:
```suggestion
    std::cout << "*** Test_MatMul: m=" << m
```

File: misc/legacy_tests/tapprox1.cpp [hunk 18]
```diff
@@ -0,0 +1,339 @@
...
+  secretKey.GenSecKey();        // A +-1/0 secret key
+  addSome1DMatrices(secretKey); // compute key-switching matrices
+
+  const PubKey& publicKey = secretKey;
+
+  if (verbose) {
+    std::cout << "security=" << context.securityLevel() << endl;
+    context.zMStar.printout();
+    cout << "r = " << context.alMod.getR() << endl;
+    cout << "ctxtPrimes=" << context.ctxtPrimes
```
Comment (esteffin) [inline misc/legacy_tests/tapprox1.cpp:319]:
```suggestion
    std::cout << "r = " << context.alMod.getR() << std::endl;
    std::cout << "ctxtPrimes=" << context.ctxtPrimes
```

File: misc/legacy_tests/tapprox1.cpp [hunk 19]
```diff
@@ -0,0 +1,339 @@
...
+	 << ", specialPrimes=" << context.specialPrimes << endl
+	 << endl;
```
Comment (esteffin) [inline misc/legacy_tests/tapprox1.cpp:321]:
```suggestion
	 << ", specialPrimes=" << context.specialPrimes << std::endl
	 << std::endl;
```

File: misc/legacy_tests/tperms1.cpp [hunk 20]
```diff
@@ -0,0 +1,113 @@
...
+#include <helib/ArgMap.h>
+
+using namespace helib;
+
+static bool noPrint = true;
+
+void testCtxt(long m, long p, long depthBound, long L, long r)
+{
+  if (!noPrint)
+    cout << "@testCtxt(m="<<m<<",p="<<p<<",depth="<<depthBound<< ",r="<<r<<")\n";
```
Comment (esteffin) [inline misc/legacy_tests/tperms1.cpp:31]:
```suggestion
    std::cout << "@testCtxt(m="<<m<<",p="<<p<<",depth="<<depthBound<< ",r="<<r<<")\n";
```

File: misc/legacy_tests/tperms1.cpp [hunk 21]
```diff
@@ -0,0 +1,113 @@
...
+
+
+  long n = context.getNSlots();
+
+  std::cout << "n=" << n << "\n";
+
+  PermIndepPrecomp pip(context, depthBound);
+
+  cout << "depth=" << pip.getDepth() << "\n";;
+  cout << "cost=" << pip.getCost() << "\n";;
```
Comment (esteffin) [inline misc/legacy_tests/tperms1.cpp:49]:
```suggestion
  std::cout << "depth=" << pip.getDepth() << "\n";;
  std::cout << "cost=" << pip.getCost() << "\n";;
```

File: misc/legacy_tests/tapprox1.cpp [hunk 22]
```diff
@@ -0,0 +1,339 @@
...
+  if (L == 0) {
+    if (R <= 2)
+      L = 100 * R;
+    else
+      L = 220 * (R - 1);
+  }
+
+  if (verbose) {
+    cout << "** m=" << m << ", #rounds=" << R << ", |q|=" << L
+         << ", epsilon=" << epsilon << endl;
```
Comment (esteffin) [inline misc/legacy_tests/tapprox1.cpp:298]:
```suggestion
    std::cout << "** m=" << m << ", #rounds=" << R << ", |q|=" << L
              << ", epsilon=" << epsilon << std::endl;
```

File: misc/legacy_tests/tapprox1.cpp [hunk 23]
```diff
@@ -0,0 +1,339 @@
...
+  buildModChain(context, L);
+
+  SecKey secretKey(context);
+  secretKey.GenSecKey();        // A +-1/0 secret key
+  addSome1DMatrices(secretKey); // compute key-switching matrices
+
+  const PubKey& publicKey = secretKey;
+
+  if (verbose) {
+    std::cout << "security=" << context.securityLevel() << endl;
```
Comment (esteffin) [inline misc/legacy_tests/tapprox1.cpp:316]:
```suggestion
    std::cout << "security=" << context.securityLevel() << std::endl;
```

File: misc/legacy_tests/Test_CKKS_matmul.cpp [hunk 24]
```diff
@@ -0,0 +1,190 @@
...
+	 << ", r=" << r
+	 << ", L=" << L
+	 << ", nt=" << nt
+	 << ", force_bsgs=" << fhe_test_force_bsgs
+	 << ", force_hoist=" << fhe_test_force_hoist
+	 << ", ks_strategy=" << ks_strategy
+	 << endl;
```
Comment (esteffin) [inline misc/legacy_tests/Test_CKKS_matmul.cpp:178]:
```suggestion
	 << std::endl;
```

File: misc/legacy_tests/tapprox1.cpp [hunk 25]
```diff
@@ -0,0 +1,339 @@
...
+  p1.encrypt(c1, /*mag=*/1.0);
+  p2.encrypt(c2, /*mag=*/1.0);
+  p3.encrypt(c3, /*mag=*/1.0);
+
+  HELIB_NTIMER_START(Circuit);
+
+  for (long i = 0; i < nRounds; i++) {
+
+    if (verbose)
+      std::cout << "*** round " << i << "..." << endl;
```
Comment (esteffin) [inline misc/legacy_tests/tapprox1.cpp:121]:
```suggestion
      std::cout << "*** round " << i << "..." << std::endl;
```

File: misc/legacy_tests/tapprox1.cpp [hunk 26]
```diff
@@ -0,0 +1,339 @@
...
+      pp3 == Approx(p3)) 
+    std::cout << "SUCCESS\n";
+  else
+    std::cout << "FAIL\n";
+
+
+  if (verbose) {
+    std::cout << endl;
+    // printAllTimers();
+    std::cout << endl;
```
Comment (esteffin) [inline misc/legacy_tests/tapprox1.cpp:250]:
```suggestion
    std::cout << std::endl;
    // printAllTimers();
    std::cout << std::endl;
```

**Review by esteffin — COMMENTED** [review]
Added full namespace qualification suggestions

File: misc/legacy_tests/Test_CKKS_matmul.cpp [hunk 27]
```diff
@@ -0,0 +1,190 @@
+/* Copyright (C) 2012-2019 IBM Corp.
```
Comment (jlhcrawford) [inline misc/legacy_tests/Test_CKKS_matmul.cpp:1]:
```suggestion
/* Copyright (C) 2020 IBM Corp.
```

File: misc/legacy_tests/tapprox1.cpp [hunk 28]
```diff
@@ -0,0 +1,338 @@
+/* Copyright (C) 2012-2020 IBM Corp.
```
Comment (jlhcrawford) [inline misc/legacy_tests/tapprox1.cpp:1]:
```suggestion
/* Copyright (C) 2020 IBM Corp.
```

File: misc/legacy_tests/tperms1.cpp [hunk 29]
```diff
@@ -0,0 +1,112 @@
+/* Copyright (C) 2012-2019 IBM Corp.
```
Comment (jlhcrawford) [inline misc/legacy_tests/tperms1.cpp:1]:
```suggestion
/* Copyright (C) 2020 IBM Corp.
```

File: src/Ptxt.cpp [hunk 30]
```diff
@@ -14,6 +14,55 @@
...
+is that for most applications, users will want to *create* PtxtArray objects
+using vectors/scalars computed by some other type of computation altogether
+(the library does this now, in fact, in various places, i.e., using linear
+algebra routines).  Once such a vector/scalar is converted to a PtxtArray, they
+can perform operations of the PtxtArray object that correspond to homomorphic
+Ctxt operations: this allows users to develop code working on ptxt objects and
+then later convert to ctxt objects with little effort (this is where Ptxt and
+PtxtArray share a similar design philosophy).
+
+A Ptxt works with the default EA/View in Context, while really, it should work
```
Comment (jlhcrawford) [inline src/Ptxt.cpp:48]:
```suggestion
```

File: src/Ptxt.cpp [hunk 31]
```diff
@@ -14,6 +14,55 @@
...
+also consistent with the way Ctxt objects are dealt with.
+
+I admit, some of my opinions are probably influenced by some form of the "not
+invented here syndrome".
+
+So for now, we have two competing classes: Ptxt and PtxtArray.  I'm not sure
+what, if anything, to do about that right now.  We will muddle forward and see
+what happens.  There are larger problems in the world.
+
+*/
```
Comment (jlhcrawford) [inline src/Ptxt.cpp:64]:
```suggestion
```

File: tests/TestCKKS.cpp [hunk 32]
```diff
@@ -334,10 +334,11 @@ TEST_P(TestCKKS, multiplyingDoubleToCiphertextWorks)
   rf = c1.getRatFactor();
   pm = c1.getPtxtMag();
   c1.multByConstantCKKS(vd[0]);
+  // c1 *= vd[0];
   ea.decrypt(c1, secretKey, vd2);
 
   mul(vd1, vd[0]);
-  rf /= std::abs(vd[0]);
+  rf /= std::abs(vd[0]); // VJS-NOTE: I fixed this with an abs
```
Comment (jlhcrawford) [inline tests/TestCKKS.cpp:341]:
```suggestion
  rf /= std::abs(vd[0]);
```

File: tests/TestMatmulCKKS.cpp [hunk 33]
```diff
@@ -0,0 +1,222 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License. See accompanying LICENSE file.
+ */
+
+/**
+ * @file matmul.h
+ * @brief some matrix / linear algenra stuff
```
Comment (jlhcrawford) [inline tests/TestMatmulCKKS.cpp:15]:
```suggestion
 * @brief some matrix / linear algebra stuff
```

**Review by jlhcrawford — APPROVED** [review]
Approved pending the passing of Travis.

**Review by faberga — APPROVED** [review]
Travis build&test passed

**Pull request merged**